include/database.py

```diff
@@ -1,18 +1,47 @@
 # database.py
 #
 # This module provides a wrapper for basic database functionality.
+import sys
 import psycopg2
 
+
 def select(connectionString, sql, parameters):
+
     # Open the connection
-    connection = psycopg2.connect(connectionString)
-    cursor = connection.cursor()
+    try:
+        connection = psycopg2.connect(connectionString, connect_timeout = 5)
+
+    # operational error is raised when parameters passed to connect() are incorrect
+    # or if the server runs out of memory
+
+    except psycopg2.OperationalError as err:
+
+        print(f' Error Connecting : {err}')
+        # set connection to 'None' in case of error
+        connection = None
+
+    # connection is successful
+    if connection is not None:
+
+        cursor = connection.cursor()
+        # catch exception for invalid SQL statement
+        try:
+            # Execute the query, note the rows
+            cursor.execute(sql, parameters)
+
+    # ProgrammingError happens when there is a syntax error in the SQL statement string passed
+    # to the psycopg2 execute() method, or if a SQL statement is executed to delete a non-existent table,
+    # or an attempt is made to create a table that already exists
+
+        except psycopg2.ProgrammingError as e:
+            print(f' Syntax Error : {e}')
+            # rollback to previous transaction before starting another
+            connection.rollback()
```

Comment:
This the structure of this try-catch is a bit more complicated than it needs to be. You can simplify things by simply wrapping the normal flow of operation in the try-catch and then having multiple `except` statements to handle the different errors.

Reply:
There are quite many errors in `psycopg2`, but I have coded for 2 errors for now.
The structure becomes a bit complex because of `if` statement along with `try` `except`.
But I have commented everything so that it looks simple and easy to understand.

Reply:
`if` statement is necessary for showing the successful connection, I guess.

Reply:
Indeed, there are quite a few exceptions possible, but don't forget that class inheritance means that you can just catch the parent exception if you don't necessary need to catch a specific exception. 

With regards to the if statement if you eliminate the second try-catch block then you don't need to test the connection since a successful connection will proceed directly to creating the cursor. A timeout would jump directly to the exception handler and bypass all of that.